cleverhans/attack_bundling.py

```diff
@@ -0,0 +1,797 @@
+    """
+    raise NotImplementedError(str(type(self)) +
+                              "needs to implement request_examples")
+
+  def is_satisfied(self, criteria, run_counts):
+    """
+    Returns a bool indicating whether the goal has been satisfied.
+    """
+    raise NotImplementedError(str(type(self)) +
+                              " needs to implement is_satisfied.")
+
+  def get_attack_config(self, attack_configs, run_counts, criteria):
+    """
+    Returns an AttackConfig to run on the next batch.
+    """
+    raise NotImplementedError(str(type(self)) +
+                              " needs to implement get_attack_config")
+
+  def new_wins(self, orig_criteria, orig_idx, new_criteria, new_idx):
+    """
+    Returns a bool indicating whether a new adversarial example is better
+    than the pre-existing one for the same clean example.
+    :param orig_criteria: dict mapping names of criteria to their value
+      for each example in the whole dataset
+    :param orig_idx: The position of the pre-existing example within the
+      whole dataset.
+    :param new_criteria: dict, like orig_criteria, but with values only
+      on the latest batch of adversarial examples
+    :param new_idx: The position of the new adversarial example within
+      the batch
+    """
+    raise NotImplementedError(str(type(self))
+                              + " needs to implement new_wins.")
+
+class Misclassify(AttackGoal):
+  """An AttackGoal that prioritizes misclassifying all examples.
+
+  Times out when each attack has been run the requested number of times.
+  Some examples may be attacked more than the goal number because we
+  always run a full batch of attacks and sometimes the batch size is
+  larger than the number of examples with attacks left to do.
+  :param new_work_goal: dict
+    Maps AttackConfigs to ints specifying how many times they should be
+    run before timing out.
+    If this dict is not specified, all attacks will be run, repeatedly,
+    until all examples are misclassified (or forever if some cannot
+    be changed into misclassified examples).
+    If this dict is specfied, only attacks in the dict will be run.
+  :param break_ties: string name of tie-breaking scheme for `new_wins`
+    When two examples are misclassified, how do we choose between them?
+    Currently the only scheme is 'wrong_confidence', where we prefer the
+    one with higher confidence assigned to a single wrong class.
+    In the future we may support other schemes like smaller perturbation
+    size, higher loss, etc.
+  """
+
+  def __init__(self, new_work_goal=None, break_ties='wrong_confidence'):
+    super(Misclassify, self).__init__()
+    self.new_work_goal = new_work_goal
+    assert all(isinstance(key, AttackConfig) for key in new_work_goal.keys())
+    assert all(isinstance(value, int) for value in new_work_goal.values())
+    self.rng = np.random.RandomState([2018, 10, 5, 9])
```

Comment:
We can make an issue but I'm not sure what the concern is. This RNG here is not particularly important. It is just used to run the attacks in a random order. Everything would still work if the attacks were run in a fixed order. The random order is just easier to implement because it doesn't require keeping track of interator state between calls to get_attack_config.

Reply:
Ok I see, then there is no need to make an issue for this then.